zrml/neo-swaps/src/lib.rs

```diff
@@ -0,0 +1,887 @@
+            Ok(Some(T::WeightInfo::exit()).into())
+        }
+
+        /// Withdraw swap fees from the specified market.
+        ///
+        /// The transaction will fail if the caller is not a liquidity provider. Should always be
+        /// used before calling `exit`.
+        ///
+        /// # Parameters
+        ///
+        /// - `market_id`: Identifier for the market related to the pool.
+        ///
+        /// # Complexity
+        ///
+        /// `O(1)`.
+        #[pallet::call_index(4)]
+        #[pallet::weight(T::WeightInfo::withdraw_fees())]
+        #[transactional]
+        pub fn withdraw_fees(
+            origin: OriginFor<T>,
+            #[pallet::compact] market_id: MarketIdOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            Self::do_withdraw_fees(who, market_id)?;
+            Ok(())
+        }
+
+        /// Deploy a pool for the specified market and provide liquidity.
+        ///
+        /// The sender specifies a vector of `spot_prices` for the market's outcomes in the order
+        /// given by the `MarketCommonsApi`. The transaction will fail if the spot prices don't add
+        /// up to exactly `BASE`.
+        ///
+        /// Depending on the values in the `spot_prices`, the transaction will transfer different
+        /// amounts of each outcome to the pool. The sender specifies a maximum `amount` of outcome
+        /// tokens to spend.
+        ///
+        /// Note that the sender must acquire the outcome tokens in a separate transaction, either by
+        /// buying from the pool or by using complete set operations. It's therefore convenient to
+        /// batch this function together with a `buy_complete_set` with `amount` as amount of
+        /// complete sets to buy.
+        ///
+        /// Deploying the pool will cost the signer an additional fee to the tune of the
+        /// collateral's existential deposit. This fee is placed in the pool account and ensures
+        /// that swap fees can be stored in the pool account without triggering dusting or failed
+        /// transfers.
+        ///
+        /// The operation is currently limited to binary and scalar markets.
+        ///
+        /// # Complexity
+        ///
+        /// `O(n)` where `n` is the number of outcomes in the specified market.
+        #[pallet::call_index(5)]
+        #[pallet::weight(T::WeightInfo::deploy_pool())]
+        #[transactional]
+        pub fn deploy_pool(
+            origin: OriginFor<T>,
+            #[pallet::compact] market_id: MarketIdOf<T>,
+            #[pallet::compact] amount: BalanceOf<T>,
+            spot_prices: Vec<BalanceOf<T>>,
+            swap_fee: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let asset_count = T::MarketCommons::market(&market_id)?.outcomes() as u32;
+            ensure!(spot_prices.len() == asset_count as usize, Error::<T>::IncorrectVecLen);
```

Comment:
This was based on the idea that benchmarking should not be mixed with business logic. Of course, that's nonsense. https://github.com/zeitgeistpm/zeitgeist/issues/1122

Reply:
In my defense, this is also defensive programming because we check if the asset count was correctly specified before doing anything else. So now I'm thinking that _in this particular case_, that's actually a good idea.